Address GitHub issues #137, #99, #98, #96, #86

laulauland opened the pull request:
This PR addresses multiple GitHub issues with improvements to the mcp-lite core library:

## Changes

### DNS Rebinding Protection Documentation (#137)
- Added documentation section in README.md explaining DNS rebinding protection
- Included code examples for configuring `allowedHosts` and `allowedOrigins`

### Multiple Binding Warning (#99)
- Added `boundServers` WeakSet to track which servers have been bound
- Added warning when binding the same HTTP transport to the same server multiple times
- Helps prevent issues with stateful servers

### JSON-RPC Error Responses for Validation (#98)
- Changed Host/Origin validation failures to return proper JSON-RPC error responses
- Returns 403 status with structured error instead of plain text

### Schema Projection Logging (#96)
- Added configurable logger option to `McpServerOptions`
- Logger defaults to stderr (console.error) for all log levels for better visibility
- Schema projection warnings are logged for dropped properties, unsupported formats, and non-string enum values
- Users can provide custom logger or disable logging entirely
- Helps developers understand why their schemas are being modified during projection

### Remove PING Notification (#86)
- Removed automatic PING notification from GET SSE stream
- Added SSE comment (`: stream ready`) to properly establish connection without sending JSON-RPC data
- Updated all integration tests to reflect removal of PING event
- Verified compatibility with SSE specification

## Testing
All 188 integration tests passing. Updated test expectations to account for removal of PING notification and logger changes.

## Closes
Closes #137
Closes #99
Closes #98
Closes #96
Closes #86

laulauland committed:
Address GitHub issues #137, #99, #98, #96, #86

- Remove .editorconfig from templates (#137)
- Warn when HTTP transport binds multiple times (#99)
- Add DNS rebinding protection docs and JSON-RPC errors (#98)
- Add schema projection logging warnings (#96)
- Replace PING with SSE comment for stream establishment (#86)

All tests passing (188 pass, 0 fail)

laulauland committed:
Add changeset for multiple issue fixes

laulauland committed:
Make logger configurable and default to stderr

laulauland committed:
Fix playground elicitation schema to use primitives instead of arrays

laulauland commented:
@codex review this please

brettimus commented on packages/core/src/context.ts:
```diff
@@ -22,6 +22,10 @@ import {
   toElicitationRequestedSchema,
 } from "./validation.js";
 
+interface Logger {
```
we already have a logger interface defined in `packages/core/src/core.ts`, this one seems like a straggler

brettimus commented on packages/core/src/transport-http/transport-http.ts:
```diff
@@ -206,7 +207,17 @@ export class StreamableHttpTransport {
+    // Check if this server has already been bound to a transport
+    if (this.boundServers.has(server)) {
+      console.warn(
+        "[mcp-lite] WARNING: Binding an HTTP transport multiple times to the same server can break stateful servers. " +
```
did biome stop complaining about string concatenation with `+`? it always hates when llms do this

brettimus commented on packages/core/src/transport-http/transport-http.ts:
```diff
@@ -206,7 +207,17 @@ export class StreamableHttpTransport {
   ) => Promise<Response> {
+    // Check if this server has already been bound to a transport
+    if (this.boundServers.has(server)) {
+      console.warn(
```
this is using `console` fwiw, not the logger

brettimus commented on packages/core/src/core.ts:
```diff
@@ -310,7 +320,13 @@ export class McpServer {
+      warn: console.error,
+      info: console.error,
+      debug: console.error,
+    };
```
i don't think we should do this. 

either no-op the noisier levels by default or switch back to "console-by-default" 

some reasoning here: `stderr` is semantically treated as errors (by downstream tooling). you don't want to treat debug logs as errors.

brettimus replied:
there is another take that libraries should be quiet by default and enable verbose logging easily, not sure how you feel about that apporach instead

brettimus commented on packages/core/src/sse-writer.ts:
```diff
@@ -45,6 +46,14 @@ export function createSSEStream(options?: { onClose?: () => void }): {
         end();
       }
     },
+    writeComment(comment: string): void {
```
how do clients treat this? do they freak out at non-jsonrpc messages? are we allowed to send just random events down the sse stream that have no bearing on the messages back and forth between client and server?

brettimus replied:
nvm i see the comment below in the code addressing this

brettimus commented on packages/core/src/transport-http/transport-http.ts:
```diff
@@ -796,11 +833,17 @@ export class StreamableHttpTransport {
     // Register the session stream
     this.sessionStreams.set(sessionId, writer);
 
+    // Send an SSE comment to establish the connection
+    // SSE streams need some initial data to properly establish the connection with the client.
+    // Without this, the client's SSE reader may not recognize the stream as "open" and will wait
+    // indefinitely for the first event. Comments (lines starting with ':') are part of the SSE spec
+    // and are ignored by clients but still flush the response and signal the stream is active.
+    // This is equivalent to Node.js's res.flushHeaders() but works in Fetch API environments.
```
good solution!

brettimus commented on packages/core/src/validation.ts:
```diff
@@ -103,9 +103,14 @@ interface ElicitationJsonSchema {
   additionalProperties?: boolean;
 }
 
+interface Logger {
```
again, feel like we could rely on the logger interface we export instead of this rogue one

brettimus commented on templates/starter-mcp-cloudflare/.editorconfig:
🗑️

brettimus reviewed (approved):
pre-approval but please fix the "everything to stderr" default, as i don't think that's a sensible default for a library to have

if anything do "quiet by default" for debug / info logs

or just revert to use console

laulauland committed:
Address PR review: make logger quiet by default (option B)

Following the adapter pattern that mcp-lite uses throughout, logging should
be opt-in rather than noisy by default. This change makes the library quiet:
- error and warn still go to console.error 
- info and debug are silent (no-op)

Users can easily enable verbose logging or provide their own logger implementation.

Also:
- Removed duplicate Logger interfaces in context.ts and validation.ts
- Added comment explaining console.warn usage in transport binding
- Added comprehensive logging documentation to README
- Updated changeset and tests

laulauland committed:
Make logger completely silent by default

Changed all log levels (error, warn, info, debug) to be no-ops by default.
This is cleaner than having error/warn go to console - libraries should be
completely silent unless the user opts in to logging.

laulauland commented:
Updated to make the logger **completely silent** by default (all levels are no-ops, including error and warn).

Now, like other parts of mcp-lite, logging follows the adapter pattern - you opt-in when you need it. I feel like this is cleaner than having error/warn go to console by default. This way the library is completely silent unless the user explicitly enables logging.

Users can easily enable logging by providing a logger:
```typescript
const mcp = new McpServer({
  name: "my-server",
  version: "1.0.0",
  logger: {
    error: console.error,
    warn: console.warn,
    info: console.info,
    debug: console.debug,
  }
});
```

Noted this in the README docs with examples for various logging configurations.

laulauland commented:
CI is failing but this is just `pkg.pr.new` bugging out
